connector/connection.go

```diff
@@ -0,0 +1,247 @@
+
+type connectionSocketState uint8
+
+const (
+	connectingConnectionSocketState connectionSocketState = iota + 1
+	connectedConnectionSocketState
+	disconnectedConnectionSocketState
+)
+
+type Connection struct {
+	state   connectionSocketState
+	stateMu sync.Mutex
+
+	loop *eventloop.EventLoop
+	*connection.Connection
+	timeout  time.Duration
+	result   chan error
+	fd       int
+	protocol connection.Protocol
+	tw       *timingwheel.TimingWheel
+	idleTime time.Duration
+	callBack connection.CallBack
+}
+
+func newConnection(
+	network, address string,
+	loop *eventloop.EventLoop,
+	timeout time.Duration,
+	protocol connection.Protocol,
+	tw *timingwheel.TimingWheel,
+	idleTime time.Duration,
+	callBack connection.CallBack) (*Connection, error) {
+
+	fd, err := unixOpenConnect(network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	connectResult := make(chan error)
+
+	conn := &Connection{
+		state:    connectingConnectionSocketState,
+		loop:     loop,
+		timeout:  timeout,
+		result:   connectResult,
+		fd:       fd,
+		protocol: protocol,
+		tw:       tw,
+		idleTime: idleTime,
+		callBack: callBack,
+	}
+
+	loop.QueueInLoop(func() {
+		if err := loop.AddSocketAndEnableRead(fd, conn); err != nil {
+			log.Info("[AddSocketAndEnableRead]", err)
+		}
+
+		if err := loop.EnableReadWrite(fd); err != nil {
+			log.Info("[EnableReadWrite] error ", err)
+		}
+	})
+
+	var (
+		ctx    context.Context
+		cancel context.CancelFunc
+	)
+
+	if timeout > 0 {
+		ctx, cancel = context.WithDeadline(context.Background(), time.Now().Add(timeout))
+		defer cancel()
+	} else {
+		ctx = context.Background()
+	}
+
+	defer close(connectResult)
+
+	select {
+	case e := <-connectResult:
+		if e != nil {
+			return nil, e
+		}
+
+		return conn, nil
+	case <-ctx.Done():
+		conn.stateMu.Lock()
+		defer conn.stateMu.Unlock()
+
+		switch conn.state {
+		case connectingConnectionSocketState:
+			conn.state = disconnectedConnectionSocketState
+			conn.closeUnconnected()
+			return nil, ErrDialTimeout
+		case connectedConnectionSocketState:
+			return conn, nil
+		default:
+			return nil, ErrDialTimeout
+		}
+	}
+}
+
+func (c *Connection) HandleEvent(fd int, events poller.Event) {
+	if c.state == connectingConnectionSocketState {
+		c.stateMu.Lock()
+		defer c.stateMu.Unlock()
+
+		if c.state != connectingConnectionSocketState {
+			return
+		}
+
+		if events&poller.EventWrite == 0 {
+			if err := checkConn(fd); err != nil {
+				c.closeUnconnected()
+				c.result <- err
+				return
+			}
+
+			sa, err := unix.Getpeername(fd)
+			if err != nil {
+				c.closeUnconnected()
+				c.result <- err
+				return
+			}
+
+			c.Connection = connection.New(c.fd, c.loop, sa, c.protocol, c.tw, c.idleTime, c.callBack)
+			c.state = connectedConnectionSocketState
+			c.result <- nil
+			c.Connection.HandleEvent(fd, events)
+			return
+		} else if events&poller.EventWrite&poller.EventErr == 0 || events&poller.EventRead&poller.EventErr == 0 {
```

Comment:
This is not a good way to deal with it. You should handle the eventErr first.

Reply:
And please ensure that all unit tests pass.
https://github.com/Allenxuxu/gev/pull/83/checks?check_run_id=2759493948#step:6:255